textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/join/JoinTest.java

```diff
@@ -0,0 +1,477 @@
+		dataWriterForOuter = new DataWriter(dataStoreForOuter, analyzer);
+		dataStoreForInner = new DataStore(DataConstants.INDEX_DIR, bookSchema);
+		dataWriterForInner = new DataWriter(dataStoreForInner, analyzer);
+		dataWriterForOuter.clearData();
+		dataWriterForInner.clearData();
+	}
+	
+	// A handy method to get join result. Called from each test case
+	public List<ITuple> getJoinResults(IOperator outer, IOperator inner, 
+			Attribute idAttribute, Attribute joinAttribute, 
+			Integer threshold) throws Exception {
+		IPredicate joinPredicate = new JoinPredicate(idAttribute, joinAttribute, 
+				threshold);
+		join = new Join(outer, inner, joinPredicate);
+		join.open();
+		
+        List<ITuple> results = new ArrayList<>();
+        ITuple nextTuple = null;
+        
+        while ((nextTuple = join.getNextTuple()) != null) {
+            results.add(nextTuple);
+        }
+        
+        return results;
+	}
+	
+	// This method cleans up after each test.
+	@After
+	public void cleanUp() throws Exception {
+		dataWriterForOuter.clearData();
+		dataWriterForInner.clearData();
+	}
+	
+	// This case tests for scenario when the IDs of the documents don't match.
+	// Test result: The list of result returned is empty.
+	@Test
+	public void testIdsDontMatch() throws Exception {
+		dataWriterForOuter.writeData(bookTuple1);
+		dataWriterForInner.writeData(bookTuple2);
+		
+		String query = "special";
+		IPredicate predicateOuter = new KeywordPredicate(query, attributeList, DataConstants.KeywordMatchingType.SUBSTRING_SCANBASED, analyzer, dataStoreForOuter);
+        keywordMatcherOuter = new KeywordMatcher(predicateOuter);
+        query = "cancer";
+        IPredicate predicateInner = new KeywordPredicate(query, attributeList, DataConstants.KeywordMatchingType.SUBSTRING_SCANBASED, analyzer, dataStoreForInner);
+        keywordMatcherInner = new KeywordMatcher(predicateInner);
+        
+        Attribute idAttr = attributeList.get(0);
+        Attribute reviewAttr = attributeList.get(4);
+        List<ITuple> resultList = getJoinResults(keywordMatcherOuter, keywordMatcherInner, idAttr, reviewAttr, 10);
+        Assert.assertEquals(0, resultList.size());
+	}
+	
+	// This case tests for the scenario when the IDs of the documents match, 
+	// fields to join match and the difference of span of terms are within 
+	// the given span threshold. 
+	// Test result: The list contains tuple with all the fields and a span 
+	// list consisting of the joined span. The joined span is made up of the 
+	// field name, start and stop index (computed as <min(span1 spanStartIndex, 
+	// span2 spanStartIndex), max(span1 spanEndIndex, span2 spanEndIndex)>) 
+	// key (combination of span1 key and span2 key) and value (combination of 
+	// span1 value and span2 value).
+	@Test
+	public void testIdsMatchFieldsMatchWithinSpan() throws Exception {
+		dataWriterForOuter.writeData(bookTuple1);
+		dataWriterForInner.writeData(bookTuple1);
+		
+		String query = "special";
```

Comment:
This piece of code is used several times in this test file.  You may optionally introduce a helper function with the necessary parameters.